src/persistent_log/fs.rs

```diff
@@ -0,0 +1,396 @@
+        // Could be more efficient about not copying data here.
+        let filelen = self.file.metadata()?.len();
+        let offset = self.offsets.get(index).ok_or(Error)?;
+        let mut buf = [0u8; 8];
+        self.file.read_exact(&mut buf)?;
+        let length = u64_from_u8s(&buf[..]) as usize;
+
+        let mut vecbuf = vec![0u8; length - 8];
+        let end = *offset as usize + length;
+        self.file.read_exact(&mut vecbuf[0..(length - 8) as usize])?;
+        let term = u64_from_u8s(&vecbuf[..8]).into();
+        let command = (&vecbuf[8..]).to_owned();
+        Ok((term, command))
+    }
+
+    fn truncate_file(&mut self, index: usize) -> Result<()> {
+        match self.offsets.get(index) {
+            None => {},
+            Some(offset) => self.file.set_len(*offset)?,
+        };
+        Ok(())
+    }
+
+    ///Add an entry to the log
+    fn write_entry(&mut self, index: usize, term: Term, command: &[u8]) -> Result<()> {
+        if index > self.entries.len() {
+            Err(Error)
+        } else {
+            let new_offset = self.file.seek(SeekFrom::End(0))?;
+            self.offsets.push(new_offset);
+            let entry_len = (command.len() + 16) as u64;
+            self.file.write_all(&u64_to_u8s(entry_len)[..])?;
+            self.file.write_all(&u64_to_u8s(term.into())[..])?;
+            self.file.write_all(&command[..])?;
+            Ok(())
+        }
+    }
+}
+
+impl Log for FsLog {
+    type Error = Error;
+
+    fn current_term(&self) -> Result<Term> {
+        Ok(self.current_term)
+    }
+
+    fn set_current_term(&mut self, term: Term) -> Result<()> {
+        self.current_term = term;
+        self.voted_for = None;
+        self.write_term()?;
+        Ok(())
+    }
+
+    fn inc_current_term(&mut self) -> Result<Term> {
+        self.current_term = self.current_term + 1;
+        self.voted_for = None;
+        self.write_term()?;
+        self.current_term()
+    }
+
+    fn voted_for(&self) -> Result<Option<ServerId>> {
+        Ok(self.voted_for)
+    }
+
+    fn set_voted_for(&mut self, address: ServerId) -> Result<()> {
+        self.voted_for = Some(address);
+        self.write_voted_for()?;
+        Ok(())
+    }
+
+    fn latest_log_index(&self) -> Result<LogIndex> {
+        Ok(LogIndex(self.entries.len() as u64))
+    }
+
+    fn latest_log_term(&self) -> Result<Term> {
+        let len = self.entries.len();
+        if len == 0 {
+            Ok(Term::from(0))
+        } else {
+            Ok(self.entries[len - 1].0)
+        }
+    }
+
+    fn entry(&self, index: LogIndex) -> Result<(Term, &[u8])> {
+        let (term, ref bytes) = self.entries[(index - 1).as_u64() as usize];
+        Ok((term, bytes))
+    }
+
+    /// Append entries sent from the leader.  
+    /// FIXME: This is contrary to the raft spec, and will cause errors, as it discards entries that may be
+    /// necessary.  The entries should not be truncated unless a mismatch is found.
```

Comment:
I think there is logic in the consensus module that discards stale messages, making this irrelevant, but it will still be useful to implement the conditional discard to limit the number of log entries that need to be rewritten in the file.